### PR TITLE
Add 7.1z1 BZ 2290711 workflow

### PR DESCRIPTION
**Pull request by sumabai**
# Description

JIRA: https://issues.redhat.com/browse/RHCEPHQE-15320
Polarion TC : https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83593402

Test Assertion : Additional clone create not started if max concurrent clone creation in-progress and snapshot_clone_no_wait is True.

Steps:
    2. Verify default value for ceph config 'ceph config get mgr mgr/volumes/snapshot_clone_no_wait' is True.
    3. Verify default value for ceph config 'ceph config get mgr mgr/volumes/max_concurrent_clones' is 4.
    4. Creats fs volume if the volume is not there
    5. Create subvolumegroup, Create subvolume in subvolumegroup and default group
    6. Run IO on subvolume, Create Snapshot.
    7. Note current max concurrent clome limit which is 4, and create 5 clones of both subvolumes in parallel across default
       and non-default subvolume groups.
    8. Verify only 4/5 clones suceed and 1 will fail with appropriate error message
       "Error EAGAIN: all cloner threads are busy, please try again later"
        Verify clone creation progress with command,
        ceph fs clone status <vol_name> <clone_name> [--group_name <group_name>]

        Use below config to simulate delay during cloning, to achieve concurrency limit,
        'ceph config set mgr mgr/volumes/snapshot_clone_delay <value_secs>'

    9. Increase max_concurrent_clones limit to 5 and Repeat step7-8 with 6 conclurrent clones.
       Verify error message on 1/6 clones.
    10. Decrease max_concurrent_clones to 3 and repeat step7-8 with 4 clones and verify error message for 1/4 clones

Logs:
passed logs with comments addressed : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-3J4URZ
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LJYYB1/clone_no_wait_if_max_concurrent_0.log
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-WZ9QLA 

Suites:
suites/reef/cephfs/tier-3_cephfs_bugs.yaml
suites/squid/cephfs/tier-3_cephfs_bugs.yaml

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

File: suites/reef/cephfs/tier-3_cephfs_bugs.yaml [hunk 1]
```diff
@@ -252,3 +252,10 @@ tests:
       polarion-id: CEPH-83591136
       desc: Verify Ceph Health Warning for Slow MDS Requests After Cluster Installation
       abort-on-fail: false
+  - test:
```
Comment (psathyan) [inline suites/reef/cephfs/tier-3_cephfs_bugs.yaml:255]: nit: ideally, I would expect this to be a separate commit. This helps in identifying changes.

File: suites/squid/cephfs/tier-3_cephfs_bugs.yaml [hunk 2]
```diff
@@ -268,4 +268,11 @@ tests:
       polarion-id: CEPH-83592513
       desc: Verify Client evict -h,--help works without running the function
       abort-on-fail: false
+  - test:
```
Comment (psathyan) [inline suites/squid/cephfs/tier-3_cephfs_bugs.yaml:271]: nit: traceability via git log would be possible if we have them in separate commits. We would also need to add better commit messages.

File: tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py [hunk 3]
```diff
@@ -0,0 +1,450 @@
+import concurrent.futures
```
Comment (psathyan) [inline tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py:1]: Please add doc strings of this test module.

File: tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py [hunk 4]
```diff
@@ -0,0 +1,450 @@
...
+from ceph.ceph import CommandFailed
+from ceph.parallel import parallel
+from tests.cephfs.cephfs_utilsV1 import FsUtils
+from tests.cephfs.cephfs_volume_management import wait_for_process
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def get_clone_status(client, fs_util, clone):
```
Comment (psathyan) [inline tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py:19]: Please docstrings for the method

File: tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py [hunk 5]
```diff
@@ -0,0 +1,450 @@
...
+            cmd_out, cmd_rc = fs_util.get_clone_status(
+                client,
+                clone["vol_name"],
+                clone["target_subvol_name"],
+                group_name=clone["target_group_name"],
+            )
+        else:
+            cmd_out, cmd_rc = fs_util.get_clone_status(
+                client, clone["vol_name"], clone["target_subvol_name"]
+            )
```
Comment (psathyan) [inline tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py:31]: we could do
```
_out, _rc = fs_util.get_clone_status(
    client,
    clone["vol_name"],
    clone["target_subvol_name"],
   group_name=clone.get("target_group_name", None),
)
```
when `get_clone_status` can perform the check for `target_group_name` as it is the interface. If it does, I would recommend making that change.

File: tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py [hunk 6]
```diff
@@ -0,0 +1,450 @@
...
+        status = json.loads(cmd_out)
+        log.info(f"Clone status for {clone['target_subvol_name']}: {status}")
+        log.info(f"{clone['target_subvol_name']} state is {status['status']['state']}")
+        return status["status"]["state"]
+    except Exception as ex:
+        log.info(ex)
+        if "does not exist" in str(ex):
+            state = "not started"
+            return state
+        else:
```
Comment (psathyan) [inline tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py:41]: you can avoid `else` like
```
except Exception as ex:
        log.info(ex)
        if "does not exist" in str(ex):
            state = "not started"
            return state
        log.error("Unknown error encountered during cloning.")
```

File: tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py [hunk 7]
```diff
@@ -0,0 +1,450 @@
...
+        out, rc = clients[0].exec_command(
+            sudo=True,
+            cmd="ceph config get mgr mgr/volumes/max_concurrent_clones",
+        )
+        if "4" not in out:
+            log.error(
+                "default value for ceph config mgr/volumes/max_concurrent_clones is not 4"
+            )
+            return 1
+
```
Comment (psathyan) [inline tests/cephfs/cephfs_bugs/clone_no_wait_if_max_concurrent.py:108]: The method tends to grow ... would it be possible to split into smaller methods

- checks
- pre-requisites
- prepare test bed
- execute
- verify
- cleanup

File: suites/reef/cephfs/tier-3_cephfs_bugs.yaml (hunk 1 above)
Comment (sumabai) [reply suites/reef/cephfs/tier-3_cephfs_bugs.yaml]: done, removed suite files, will include it another PR

File: suites/squid/cephfs/tier-3_cephfs_bugs.yaml (hunk 2 above)
Comment (sumabai) [reply suites/squid/cephfs/tier-3_cephfs_bugs.yaml]: done, removed suite files, will include it another PR

**Review by psathyan — COMMENTED** [review]
/lgtm

**Review by hkadam134 — APPROVED** [review]
lgtm

**Review by AmarnatReddy — APPROVED** [review]
LGTM

**Comment by psathyan** [comment]
please rebase your changes for the new gate check `tox (3.9.18)` to be executed

**Review by psathyan — APPROVED** [review]
/lgtm
